## Exploring Ebay Car sales Data
In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.
The aim of this project is to clean the data and analyze the included used car listings.



In [2]:
import numpy as np
import pandas as pd

autos = pd.read_csv("autos.csv",encoding = "Latin-1")

Let us create a new cell and print only autos.The feature of jypter notebooks is its ability to render the first few and last few values of any pandas object

In [3]:
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

we can make the following observations from the above output:

* The dataset contains 20 columns, most of which are strings.
* Some columns have null values, but none have more than ~20% null values.
* The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [5]:
#to print the array of existing column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We are going to rename the column names in the dataframes

make the following edits to columns names:
* yearOfRegistration to registration_year
* monthOfRegistration to registration_month
* notRepairedDamage to unrepaired_damage
* dateCreated to ad_created
* The rest of the columnn names from camelcase to snakecase.

In [6]:
#to rename the column names we are going to use rename function

autos.rename(columns = {'yearOfRegistration':'registration_year','monthOfRegistration':'registration_month',
                        'notRepairedDamage':'unrepaired_damage','dateCreated':'ad_created',
                        'dateCrawled':'date_crawled','offerType':'offer_type','vehicleType':'vehicle_type',
                       'powerPS':'power_ps','fuelType':'fuel_type','nrOfPictures':'nr_of_pictures',
                        'postalCode':'postal_code','lastSeen':'last_seen'}
             ,inplace = True)



autos.columns
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [7]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN


Our initial observations:

* There are a number of text columns where all (or nearly all) of the values are the same:
  * seller
  * offer_type
* The num_photos column looks odd, we'll need to investigate this further.

In [8]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

There is no data in the nr_of_pictures hence we are dropping it along with seller and offer_type

In [9]:
autos = autos.drop(['nr_of_pictures','seller','offer_type'],axis = 1)

In [10]:
autos.describe(include= 'all')

date_crawled         name  price abtest vehicle_type  \
count                 50000        50000  50000  50000        44905   
unique                48213        38754   2357      2            8   
top     2016-03-30 17:37:35  Ford_Fiesta     $0   test    limousine   
freq                      3           78   1421  25756        12859   
mean                    NaN          NaN    NaN    NaN          NaN   
std                     NaN          NaN    NaN    NaN          NaN   
min                     NaN          NaN    NaN    NaN          NaN   
25%                     NaN          NaN    NaN    NaN          NaN   
50%                     NaN          NaN    NaN    NaN          NaN   
75%                     NaN          NaN    NaN    NaN          NaN   
max                     NaN          NaN    NaN    NaN          NaN   

        registration_year  gearbox      power_ps  model   odometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN  manuell           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

The price and odometer columns are numeric values stored as text. For each column:
* Remove any non-numeric characters.
* Convert the column to a numeric dtype.
* Use DataFrame.rename() to rename the column to odometer_km.

In [11]:
autos["price"] = autos["price"].str.replace('$','').str.replace(',','').astype(int)

In [12]:
#converting the odometer column from string to int

autos["odometer"] = autos["odometer"].str.replace('km','').str.replace(',','').astype(int)
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

### Exploration of odometer_km and price columns
we are going to analysis further odometer_km and price columns. there might be some wrong value in it.

In [13]:
print(autos["price"].unique().shape)

(2357,)


In [14]:
print(autos["price"].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [15]:
print(autos["price"].value_counts().head(20))

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64


From the above output, we could see 1421 rows is having 0 value in the price. So we are going to drop these rows in dataset.

In [16]:
#to view max value in the price columns
print(autos["price"].value_counts().sort_index(ascending= False).head(20))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


In [17]:
#to view min value in the price columns
print(autos["price"].value_counts().sort_index(ascending= True).head(20))

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [18]:
autos = autos[autos["price"].between(1,350000)]
print(autos["price"].describe())

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64


In [19]:
#Exploration of odometer_km columns

print(autos["odometer_km"].unique().shape)

(13,)


In [20]:
print(autos["odometer_km"].describe())

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [21]:
print(autos["odometer_km"].value_counts().head(20))

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64


The values for the odometer_km seems good. so we can start analysing the other columns.

## Exploring the date columns
There are a number of columns with date information:

* date_crawled
* registration_month
* registration_year
* ad_created
* last_seen

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.